RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/code_models/Fragment.java

```diff
@@ -0,0 +1,394 @@
+            features.add(new CSVItem(Feature.TotalSymbolsInCodeFragment, getBody().length()));
+            features.add(new CSVItem(Feature.AverageSymbolsInCodeLine, (double) getBody().length() / getBodyLineLength()));
+            features.add(new CSVItem(Feature.TotalLinesOfCode, getBodyLineLength()));
+        }
+
+        private void rankingScoreComputation(double lengthSensitivity, double maxLengthScore) {
+            RankEvaluater ranker = new RankEvaluater(this, lengthSensitivity, maxLengthScore);
+            this.score = ranker.getScore();
+        }
+
+        int analyzeDepth(String code) {
+            int dep = 0;
+            int area = 0;
+            int depInLine = 0;
+            for (Character ch : code.toCharArray()) {
+                if (ch == '{') {
+                    dep++;
+                    depInLine++;
+                } else if (ch == '}') {
+                    dep--;
+                    depInLine--;
+                } else if (ch == '\n') {
+                    int resDep = dep;
+                    if (depInLine > 0) {
+                        resDep--;
+                    }
+                    depInLine = 0;
+                    area += resDep;
+                }
+            }
+            return area;
+        }
+
+        private int getBodyLineLength() {
+            return (StringUtils.countMatches(getBody(), '\n') + 1);
+        }
+
+        private int getMethodLineLength() {
+            return (StringUtils.countMatches(getMethod(), '\n') + 1);
+        }
+
+        /**
+         * Returns body of the fragment (sequence of statements)
+         * w/o leading and trailing curly braces, whitespaces, linebreaks
+         */
+        private String getBody() {
+            return clearCode(this.methodDeclaration.getBody().toString());
+        }
+
+        private String getMethod() {
+            return this.methodDeclaration.toString();
+        }
+
+        private void writeFeatures(FileWriter fw) throws IOException {
+            int nFeatures = features.size();
+            features.sort(Comparator.comparingInt(ICSVItem::getId));
+            for (int i = 0; i < nFeatures; ++i) {
+                fw.append(String.valueOf(features.get(i).getValue()));
+                fw.append(';');
+            }
+            fw.append(String.valueOf(this.score)); //Score with 4 decimal places
+            fw.append(";");
+            fw.append("-\n");
+        }
+
+        private void process(FileWriter fw) throws IOException {
+            keywordFeaturesComputation();
+            historicalFeaturesComputation();
+            couplingFeaturesComputation();
+            methodDeclarationFeaturesComputation();
+            lengthFeaturesComputation();
+
+            double lengthScoreSensitivity = 0.1;
+            double maxLengthScore = 4;
+
+            rankingScoreComputation(lengthScoreSensitivity, maxLengthScore);
+
+            writeFeatures(fw);
+        }
+
+        protected class RankEvaluater {
```

Comment:
in my view, it would be better to extract all stuff that relates to Has into separate class, not store it as an inner class here

Reply:
will look at it